Issue #357: Use ConfigReplace when full config is sent over gNMI

anders-nexthop opened the pull request:
<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it

By default, gNMI does a version of `config reload` when passed a full config. This is disruptive, since it requires services to be restarted for it to take effect. It's also not intuitive; all of the other SET gnmi commands take place immediately. By enhancing the full config endpoint handling to trigger `config replace` instead of the gnmi-specific version of reload that exists today, we resolve both of those issues.

The existing implementation uses a full config DELETE given alongside a full config UPDATE to represent a reload. We can leave that syntax and behavior intact and instead add logic to handle a full config REPLACE.

fixes [Issue 357](https://github.com/sonic-net/sonic-gnmi/issues/357)

#### How I did it

There is already a D-BUS endpoint implemented for `config replace`, so all this takes is a handler implemented in [gnmi/sonic_service_client/dbus_client.go](https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_service_client/dbus_client.go) which will pass the full config it receives through to HostServices (and from there to GCU).

#### How to verify it

Manually verified by sending full configs using `gnmic` and ensuring that state changes are propagated to ConfigDb.

DBUS message passing verified by using `dbus-monitor` to verify content of messages being passed.

All sonic-gnmi unit tests pass. New test cases were added to verify the SET-REPLACE handler logic. A new unit test was added to `dbus_client_test.go` to test the new D-Bus endpoint.

A sonic-mgmt testcase for this behavior is proposed [here](https://github.com/sonic-net/sonic-mgmt/issues/17885).

#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Use ConfigReplace when full config is sent over gNMI

#### Link to config_db schema for YANG module changes
<!--
Provide a link to config_db schema for the table for which YANG model
is defined
Link should point to correct section on https://github.com/Azure/SONiC/wiki/Configuration.
-->

#### A picture of a cute animal (not mandatory but encouraged)

   _
 / . . \                        
  \   /.   \__________/ \
    |    \--------------/
    ^

(be kind i drew it myself)

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

qiluo-msft commented on sonic_data_client/mixed_db_client.go:
```diff
@@ -1385,19 +1385,23 @@ func (c *MixedDbClient) SetFullConfig(delete []*gnmipb.Path, replace []*gnmipb.U
 	}
 	content := []byte(ietf_json_val)
+
+	// still needed...?
```
>// still needed...? [](http://example.com/codeflow?start=1&length=19)

This is a good question. I guess ConfigReplace already validating running config?

anders-nexthop replied:
Yes, ConfigReplace does that validation. It also ensures atomicity, so no config changes will be made unless the whole config is valid.

I left this in to preserve the existing behavior, at least to some extent. So on a gNOI Reboot, the config passed here will be picked up. But I'm not sure if we want to keep that behavior or not.

ganglyu replied:
We can remove the yang validation

anders-nexthop replied:
I will remove the temp file.

I also see that in SetIncrementalConfig() above, we call ConfigSave() if the configuration is successfully accepted, so that `/etc/sonic/config_db.json` is updated. Should we do that here as well

anders-nexthop replied:
I added a call to ConfigSave() after the ConfigReload() completes successfully to make this command persistent.

ganglyu commented:
@anders-nexthop 
https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_data_client/mixed_db_client.go#L1454-L1463
For now, we use special syntax for a full configuration update. When we switch to using config replace, we should also change the syntax to replace "/".

anders-nexthop commented:
> @anders-nexthop https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_data_client/mixed_db_client.go#L1454-L1463 For now, we use special syntax for a full configuration update. When we switch to using config replace, we should also change the syntax to replace "/".

I don't quite follow. I thought a full config update was represented with both and empty 'delete' path and an empty 'update' path. Is that not correct? What would we do differently to represent a config replace? If we can represent this operation in a different way we can leave the existing implementation in place, which may be desirable for compatibility. Are you perhaps suggesting that we could instead use the 'replace' path, passing in an empty path with the file contents as a value? For example:
```
    # from sonic-mgmt/tests/gnmi/test_gnmi_configdb.py
    delete_list = ["/sonic-db:CONFIG_DB/localhost/"]
    update_list = ["/sonic-db:CONFIG_DB/localhost/:@/root/%s" % filename]
    replace_list = ["/sonic-db:CONFIG_DB/localhost/:@/root/%s" % filename]

    gnmi_set(duthost, ptfhost, delete_list, update_list, [])                                       # this is the existing syntax

    gnmi_set(duthost, ptfhost, [], [], replace_list)                                                     # this would trigger ConfigReplace
```

ganglyu commented:
@anders-nexthop 
yes, you are correct, we could instead use the 'replace' path, passing in an empty path with the file contents as a value

anders-nexthop commented:
> @anders-nexthop yes, you are correct, we could instead use the 'replace' path, passing in an empty path with the file contents as a value

Yes that makes sense. I will make that change and update the PR.

ganglyu commented:
yes, this operation should be persistent

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

qiluo-msft commented:
> > > @anders-nexthop https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_data_client/mixed_db_client.go#L1454-L1463 For now, we use special syntax for a full configuration update. When we switch to using config replace, we should also change the syntax to replace "/".
> > 
> > 
> > I don't quite follow. I thought a full config update was represented with both and empty 'delete' path and an empty 'update' path. Is that not correct? What would we do differently to represent a config replace? If we can represent this operation in a different way we can leave the existing implementation in place, which may be desirable for compatibility. Are you perhaps suggesting that we could instead use the 'replace' path, passing in an empty path with the file contents as a value? For example:
> > ```
> >     # from sonic-mgmt/tests/gnmi/test_gnmi_configdb.py
> >     delete_list = ["/sonic-db:CONFIG_DB/localhost/"]
> >     update_list = ["/sonic-db:CONFIG_DB/localhost/:@/root/%s" % filename]
> >     replace_list = ["/sonic-db:CONFIG_DB/localhost/:@/root/%s" % filename]
> > 
> >     gnmi_set(duthost, ptfhost, delete_list, update_list, [])                                       # this is the existing syntax
> > 
> >     gnmi_set(duthost, ptfhost, [], [], replace_list)                                                     # this would trigger ConfigReplace
> > ```
> 
> @ganglyu @qiluo-msft I have updated the PR to be based off of the `replace_list`, which keeps the original behavior intact if using the the delete_list/update_list syntax. The build is failing with an issue that seems unrelated:
> 
> ```
> error: sonic-interface:29: module "sonic-vnet" not found in search path
> ```
> 
> I added new test cases to cover this change. They are passing locally but do to the build problem referenced above they have not been run through the pipeline yet.

Seems like the build failure is resolved by https://github.com/sonic-net/sonic-buildimage/pull/22293

mssonicbld commented:
/azp run

anders-nexthop commented:
> > > > @anders-nexthop https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_data_client/mixed_db_client.go#L1454-L1463 For now, we use special syntax for a full configuration update. When we switch to using config replace, we should also change the syntax to replace "/".
> > > 
> > > 
> > > I don't quite follow. I thought a full config update was represented with both and empty 'delete' path and an empty 'update' path. Is that not correct? What would we do differently to represent a config replace? If we can represent this operation in a different way we can leave the existing implementation in place, which may be desirable for compatibility. Are you perhaps suggesting that we could instead use the 'replace' path, passing in an empty path with the file contents as a value? For example:
> > > ```
> > >     # from sonic-mgmt/tests/gnmi/test_gnmi_configdb.py
> > >     delete_list = ["/sonic-db:CONFIG_DB/localhost/"]
> > >     update_list = ["/sonic-db:CONFIG_DB/localhost/:@/root/%s" % filename]
> > >     replace_list = ["/sonic-db:CONFIG_DB/localhost/:@/root/%s" % filename]
> > > 
> > >     gnmi_set(duthost, ptfhost, delete_list, update_list, [])                                       # this is the existing syntax
> > > 
> > >     gnmi_set(duthost, ptfhost, [], [], replace_list)                                                     # this would trigger ConfigReplace
> > > ```
> > 
> > 
> > @ganglyu @qiluo-msft I have updated the PR to be based off of the `replace_list`, which keeps the original behavior intact if using the the delete_list/update_list syntax. The build is failing with an issue that seems unrelated:
> > ```
> > error: sonic-interface:29: module "sonic-vnet" not found in search path
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I added new test cases to cover this change. They are passing locally but do to the build problem referenced above they have not been run through the pipeline yet.
> 
> Seems like the build failure is resolved by [sonic-net/sonic-buildimage#22293](https://github.com/sonic-net/sonic-buildimage/pull/22293)

@qiluo-msft the build failed again with a similar failure:

[error: /usr/local/cvlyang-models/sonic-vnet.yang:18: module "sonic-vxlan" not found in search path
](https://dev.azure.com/mssonic/build/_build/results?buildId=823423&view=logs&j=83516c17-6666-5250-abde-63983ce72a49&t=335d3395-6d29-5762-4daa-c188603f3eb8&l=267)

mssonicbld commented:
/azp run

anders-nexthop commented:
@qiluo-msft and @ganglyu I believe I have responded to all of the outstanding review comments, can I get approval for this change?

mssonicbld commented:
/azp run
